gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -312,34 +312,88 @@ for _, ply in pairs(player.GetHumans()) do
     DrpCanHear[ply] = {}
 end
 
--- proxy function to take load from PlayerCanHearPlayersVoice, which is called a quadratic amount of times per tick,
--- causing a lagfest when there are many players
-local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    DrpCanHear[listener] = DrpCanHear[listener] or {}
-    if listener:IsBot() then return end
-    local shootPos = listener:GetShootPos()
-    for _, talker in ipairs(player.GetHumans()) do
-        local talkerShootPos = talker:GetShootPos()
-        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
-            (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
-                (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
-    end
-end
-hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
-    calcPlyCanHearPlayerVoice(ply)
-    if ply:IsBot() then return end
-    timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
-end)
+local gridSize = GM.Config.voiceDistance*2 -- Grid cell size is equal to the size of the diamater of player talking
+local floor = math.floor -- Caching floor as we will need to use it a lot
+
+-- Grid based position check
+timer.Create("DarkRPCanHearPlayersVoice", 1, 0, function()
+    DrpCanHear = {}
+    local players = player.GetHumans()
+
+    -- Create Grid
+    local grid = {}
+    local plyToGrid = {
+        {},
+        {}
+    }
+
+    -- Get the grid position of every player O(N)
+    for _, ply in ipairs(players) do
+        local pos = ply:GetShootPos()
+        local x = floor(pos.x/gridSize)
+        local y = floor(pos.y/gridSize)
+        grid[x] = grid[x] or {}
+        grid[x][y] = grid[x][y] or {}
+        grid[x][y][#grid[x][y] + 1] = ply
+        plyToGrid[1][ply] = x
+        plyToGrid[2][ply] = y
+
+        DrpCanHear[ply] = {} -- Initialize output variable
```

Comment:
According to [this stackoverflow suggestion](https://stackoverflow.com/questions/16239563/is-it-better-to-set-a-table-to-empty-or-set-all-elements-of-a-table-to-nil) it is better to clear keys the keys. I am not sure if there is a function, but I am happy to loop over all keys with `pairs`

Reply:
There is [`table.Empty`](https://wiki.facepunch.com/gmod/table.Empty). I thought it was a LuaJIT primop, but it's just a for loop.

Im not sure if it's better or worse in this specific situation. It might need some measurements.

Reply:
It also mentions that this is only true when there are 32+ entries, which I think is unlikely for a single person to check against 32 other people.

Reply:
table.Empty loops thru the table to empty it, it's faster to recreate the table.